File: 704-binary-search/704-binary-search.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
    let beg = 0, end = nums.length - 1
    while (beg <= end) {
        mid = beg + ~~((end - beg) / 2)
        if (nums[mid] == target)
            return mid;
        else if (nums[mid] < target) {
            beg = mid + 1
        }
        else {
            end = mid - 1
        }
    }
    return -1;
};

File: 704-binary-search/704-binary-search.py
class Solution:
    def search(self, arr: List[int], key: int) -> int:
        beg, end = 0, len(arr) - 1
        found = False
        while beg <= end:
            mid = beg + (end - beg) // 2	# to avoid mid overflow
            if key == arr[mid]:				# condition
                found = True
                return mid	 				# found index
            elif key < arr[mid]:
                end = mid - 1
            else:
                beg = mid + 1
        if not found:
            return -1